File: C++/Greedy/1931.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <algorithm>

#define LMIT 100001

typedef struct _Time
{
    int start;
    int end;
} Time;

bool comp(Time &t1, Time &t2)
{
    if (t1.end == t2.end)
        return t1.start < t2.start; //종료시간이 같을 경우 시작시간이 작은 순
    else
        return t1.end < t2.end; //아니면 종료시간이 작을 때
}

using namespace std;
int main()
{
    int n = 0;
    scanf("%d", &n);
    vector<Time> v(n);
    for (int i = 0; i < n; i++)
    {
        scanf("%d %d", &v[i].start, &v[i].end);
    }
    sort(v.begin(), v.end(), comp); //먼저 정렬함
    int endTime = 0;                //바로 직전 회의 종료시간
    int cnt = 0;
    for (int i = 0; i < v.size(); i++)
    {
        if (endTime <= v[i].start) //다음 회의 시작시간이 바로 직전 종료시간보다 크거나 같으면 회의가 가능
        {
            endTime = v[i].end; //종료시간 변경
            cnt++;              //증가
        }
    }
    printf("%d", cnt);
}

File: Go/Greedy/1931.go
package main

import (
	"fmt"
	"sort"
)

type Time struct {
	start int
	end   int
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	N := 0
	fmt.Scan(&N)
	grd := make([]Time, N)
	for i := 0; i < N; i++ {
		fmt.Scan(&grd[i].start, &grd[i].end)
	}
	sort.Slice(grd, func(i, j int) bool {
		if grd[i].end == grd[j].end {
			return grd[i].start < grd[j].start
		} else {
			return grd[i].end < grd[j].end
		}
	})
	cnt := 1
	endTime := grd[0].end
	for i := 1; i < N; i++ {
		if endTime <= grd[i].start {
			endTime = grd[i].end
			cnt++
		}
	}
	fmt.Println(cnt)

}
